docs/indexing.md

```diff
@@ -0,0 +1,422 @@
+
+#### Example
+
+For a broadcast from `tensor<20xf32>` to `tensor<10x20x30xf32>`
+
+```c
+p0 = f32[20] parameter(0)
+bc0 = f32[10, 20, 30] broadcast(p0), dimensions={1}
+```
+
+the indexing map from the output to input is (i, j, k) \mapsto (j)$ for i ∈ [0, 10), j ∈ [0, 20) and k ∈ [0, 30). 
+
+## Motivation
+
+There are many parts of XLA, which would benefit from knowledge about indexing.
+
+### **Coalescing**
+
+Detecting coalesced memory access without knowing what elements/slices of the inputs are read to compute an element of the output becomes possible only in the simplest cases.
+
+### **Utilization**
+
+Operand utilization in XLA indicates how much each input of the instruction is used assuming its output is fully used. Currently, utilization is also not computed for a generic case. Indexing analysis allows to compute utilization precisely.
+
+### **Tiling**
+
+Tiles are parameterized by offsets, sizes and strides. Tile propagation is a way to compute tile parameters of the producer/consumer of the op using the tiling parameters of the op itself. There is already a [library](https://github.com/openxla/xla/blob/main/xla/service/gpu/triton_tiling_propagation.h) that does it for softmax and dot. Tile propagation can be made more generic and robust if it is expressed via indexing maps.
+
+## **Function and Domain**
+
+The indexing map is a function $\boldsymbol{f}(\boldsymbol{d}, \boldsymbol{s})$ that maps a multi-index  $\boldsymbol{d}$ of a tensor $A$ to elements/ranges of tensor $B$. The parameter $\boldsymbol{s}$ refers to the ranges of indices of  the dimensions that are present in tensor $B$, but not in tensor $A$​.
+
+For example, if we have a reduction from `tensor<2x4x8x16xf32>` to `tensor<4x8xf32>`, then the indexing map from the 2D output to the 4D input is $(d_0, d_1) \mapsto (s_0, d_0, d_1, s_1)$, where $d_i$ are the dimension parameters that correspond to the indices of the output tensor. Parameters $s_j$ encode multiple values, i.e. to compute a $(d_0, d_1)$ element of the output, we need $(s_0, d_0, d_1, s_1)$ elements of the input, where $s_0 \in [0, 2)$ and $s_1 \in [0, 16)$. 
+
+This mapping can be constructed from the attributes of HLO instructions or the mappings of unfused instructions can be composed to get indexing for a fusion. The mapping also has a domain, which specifies for what elements of the tensor the mapping exists. 
+$$
+\begin{eqnarray}
+\boldsymbol{f}(\boldsymbol{d}, \boldsymbol{s})\; &s.t.& \\
+ \boldsymbol{lb}_d &\leq& \boldsymbol{d} \leq \boldsymbol{ub}_d \\
+  \boldsymbol{lb}_s &\leq& \boldsymbol{s} \leq \boldsymbol{ub}_s \\
+   \boldsymbol{lb}_g &\leq& \boldsymbol{g}(\boldsymbol{d}, \boldsymbol{s}) \leq \boldsymbol{ub}_g
+\end{eqnarray}
+$$
```

Comment:
This is not rendered correctly in the preview:
https://github.com/openxla/xla/blob/fbd61ef93e8613c54a0273d14e78ff46aca99edf/docs/indexing.md